Firestore/core/test/firebase/firestore/remote/grpc_stream_test.cc

```diff
@@ -0,0 +1,318 @@
+  EXPECT_NO_THROW(stream->Finish());
+}
+
+TEST_F(GrpcStreamTest, CanFinishAfterStarting) {
+  stream->Start();
+  EXPECT_NO_THROW(stream->Finish());
+}
+
+TEST_F(GrpcStreamTest, CannotFinishTwice) {
+  stream->Start();
+  EXPECT_NO_THROW(stream->Finish());
+  EXPECT_ANY_THROW(stream->Finish());
+}
+
+TEST_F(GrpcStreamTest, CanWriteAndFinishBeforeStarting) {
+  EXPECT_NO_THROW(stream->WriteAndFinish({}));
+}
+
+TEST_F(GrpcStreamTest, CanWriteAndFinishAfterStarting) {
+  stream->Start();
+  EXPECT_NO_THROW(stream->WriteAndFinish({}));
+}
+
+TEST_F(GrpcStreamTest, ObserverReceivesOnStart) {
+  stream->Start();
+  ForceFinish({/*Start*/ Ok});
+  EXPECT_EQ(observer->observed_states, States({"OnStreamStart"}));
+}
+
+TEST_F(GrpcStreamTest, CanWriteAfterStreamIsOpen) {
+  stream->Start();
+  ForceFinish({/*Start*/ Ok});
+  EXPECT_NO_THROW(stream->Write({}));
+}
+
+TEST_F(GrpcStreamTest, ObserverReceivesOnRead) {
+  stream->Start();
+  ForceFinish({/*Start*/ Ok, /*Read*/ Ok});
+  EXPECT_EQ(observer->observed_states,
+            States({"OnStreamStart", "OnStreamRead"}));
+}
+
+TEST_F(GrpcStreamTest, ReadIsAutomaticallyReadded) {
+  stream->Start();
+  ForceFinish({/*Start*/ Ok, /*Read*/ Ok});
+  EXPECT_EQ(observer->observed_states,
+            States({"OnStreamStart", "OnStreamRead"}));
+
+  ForceFinish({/*Read*/ Ok});
+  EXPECT_EQ(observer->observed_states,
+            States({"OnStreamStart", "OnStreamRead", "OnStreamRead"}));
+}
+
+TEST_F(GrpcStreamTest, ObserverReceivesOnWrite) {
+  stream->Start();
+  ForceFinish({/*Start*/ Ok});
+
+  stream->Write({});
+  ForceFinish({/*Read*/ Ignore, /*Write*/ Ok});
+
+  EXPECT_EQ(observer->observed_states,
+            States({"OnStreamStart", "OnStreamWrite"}));
+}
+
+TEST_F(GrpcStreamTest, CanAddSeveralWrites) {
+  stream->Start();
+  ForceFinish({/*Start*/ Ok});
+
+  stream->Write({});
+  stream->Write({});
+  stream->Write({});
+  ForceFinish({/*Read*/ Ignore, /*Write*/ Ok, /*Write*/ Ok, /*Write*/ Ok});
+
+  EXPECT_EQ(observer->observed_states,
+            States({"OnStreamStart", "OnStreamWrite", "OnStreamWrite",
+                    "OnStreamWrite"}));
+}
+
+TEST_F(GrpcStreamTest, ObserverReceivesOnError) {
+  stream->Start();
+  ForceFinish({/*Start*/ Ok});
+
+  // Fail the read, but allow Finish to succeed
+  ForceFinish({/*Read*/ Error, /*Finish*/ Ok});
+  EXPECT_EQ(observer->observed_states,
+            States({"OnStreamStart", "OnStreamError"}));
+}
+
+TEST_F(GrpcStreamTest, ObserverDoesNotReceiveOnFinishIfCalledByClient) {
+  stream->Start();
+  ForceFinish({/*Start*/ Ok});
+
+  stream->Finish();
+  ForceFinish({/*Read*/ Ignore, /*Finish*/ Ok});
+  EXPECT_EQ(observer->observed_states, States({"OnStreamStart"}));
+}
+
+TEST_F(GrpcStreamTest, WriteAndFinish) {
+  stream->Start();
+  ForceFinish({/*Start*/ Ok});
+
+  stream->WriteAndFinish({});
+  ForceFinish({/*Read*/ Ignore, /*Write*/ Ok, /*Finish*/ Ok});
+  // Should be no notification on the final write.
```

Comment:
I know that the on-write callback is optional; just trying to understand the design. Is it that all except the last write will trigger on-write callback? If so, what makes the final write special?

Reply:
I removed `OnWrite` callback altogether, because Firestore stream never actually cares about it.